crates/node/src/lib.rs

```diff
@@ -70,56 +73,61 @@ pub enum BuildError {
         #[code]
         worker::WorkerRoleBuildError,
     ),
-    #[error("invalid controller endpoint: {0}")]
+    #[error("node neither runs cluster controller nor its address has been configured")]
     #[code(unknown)]
-    InvalidControllerEndpoint(#[from] InvalidUri),
+    UnknownClusterController,
 }
 
 pub struct Node {
-    node_id: NodeId,
-    cluster_controller_endpoint: Uri,
+    node_id: PlainNodeId,
+    cluster_controller_address: NetworkAddress,
 
     cluster_controller_role: Option<ClusterControllerRole>,
-    worker_role: WorkerRole,
+    worker_role: Option<WorkerRole>,
     node_ctrl: NodeCtrlService,
 }
 
 impl Node {
-    pub fn new(
-        node_id: impl Into<NodeId>,
-        cluster_controller_location: ClusterControllerLocation,
-        options: Options,
-    ) -> Result<Self, BuildError> {
-        let cluster_controller_role = if let ClusterControllerLocation::Local =
-            cluster_controller_location
-        {
+    pub fn new(options: Options) -> Result<Self, BuildError> {
+        let cluster_controller_role = if options.roles.contains(Role::ClusterController) {
             Some(ClusterControllerRole::try_from(options.clone()).expect("should be infallible"))
         } else {
             None
         };
 
-        let worker_role = WorkerRole::try_from(options.clone())?;
+        let worker_role = if options.roles.contains(Role::Worker) {
+            Some(WorkerRole::try_from(options.clone())?)
+        } else {
+            None
+        };
 
         let node_ctrl = options.node_ctrl.build(
-            Some(worker_role.rocksdb_storage().clone()),
-            worker_role.bifrost_handle(),
+            worker_role
+                .as_ref()
+                .map(|worker| (worker.rocksdb_storage().clone(), worker.bifrost_handle())),
             cluster_controller_role
                 .as_ref()
                 .map(|cluster_controller| cluster_controller.handle()),
         );
 
-        let cluster_controller_endpoint =
-            if let ClusterControllerLocation::Remote(cluster_controller_address) =
-                cluster_controller_location
-            {
-                cluster_controller_address.parse()?
-            } else {
-                node_ctrl.endpoint()
-            };
+        let cluster_controller_address = if let Some(cluster_controller_address) =
+            options.cluster_controller_address
+        {
+            if cluster_controller_role.is_some() {
+                warn!("This node is running the cluster controller but has also a remote cluster controller address configured. \
+                This indicates a potential misconfiguration. Trying to connect to the remote cluster controller.");
+            }
+
+            cluster_controller_address
+        } else if cluster_controller_role.is_some() {
+            NetworkAddress::DnsName(format!("localhost:{}", node_ctrl.port()))
```

Comment:
```suggestion
            NetworkAddress::DnsName(format!("127.0.0.1:{}", node_ctrl.port()))
```

Nit: Avoid ns lookups whenever possible.